File: last_number.cpp
#include "last_number.h"

#include <Arduino.h>
#include <EEPROM.h>
#include <string.h>

/* Keep in sync with tools/pulse_monitor/last_number.py */
#define LAST_NUMBER_MAGIC 0x4E
#define LAST_NUMBER_MAX_LEN 29

/*
 * chg_hist owns 16..78 (base 16, 3 header bytes + 12 * 5-byte records).
 * Start clear of it; the ATmega2560 has 4096 bytes of EEPROM.
 */
#define LAST_NUMBER_EE_BASE 80
#define LAST_NUMBER_OFF_MAGIC (LAST_NUMBER_EE_BASE + 0)
#define LAST_NUMBER_OFF_LEN (LAST_NUMBER_EE_BASE + 1)
#define LAST_NUMBER_OFF_CHARS (LAST_NUMBER_EE_BASE + 2)


/*
 * Length of the stored number, or 0 when there is no usable record.
 *
 * Both checks are load-bearing. Erased EEPROM reads as 0xFF everywhere and a
 * zeroed one as 0x00, so without the magic a board that has never placed a
 * call would offer to redial a number made of 0xFF bytes. Without the length
 * bound a corrupt record would read past itself and return whatever follows
 * it in EEPROM.
 */
static uint8_t last_number_stored_len(void)
{
	if (EEPROM.read(LAST_NUMBER_OFF_MAGIC) != LAST_NUMBER_MAGIC)
		return 0;

	uint8_t len = EEPROM.read(LAST_NUMBER_OFF_LEN);
	if (len < 1 || len > LAST_NUMBER_MAX_LEN)
		return 0;

	return len;
}


void last_number_store(const char *number)
{
	if (number == nullptr || number[0] == '\0')
		return;

	uint8_t len = (uint8_t)strnlen(number, LAST_NUMBER_MAX_LEN);
	if (len == 0)
		return;

	/*
	 * update() only touches bytes that changed, so redialling the same
	 * number repeatedly costs no EEPROM write cycles.
	 */
	EEPROM.update(LAST_NUMBER_OFF_MAGIC, LAST_NUMBER_MAGIC);
	EEPROM.update(LAST_NUMBER_OFF_LEN, len);
	for (uint8_t i = 0; i < len; i++)
		EEPROM.update(LAST_NUMBER_OFF_CHARS + i, (uint8_t)number[i]);
}


bool last_number_available(void)
{
	return last_number_stored_len() > 0;
}


bool last_number_load(char *out, uint8_t out_len)
{
	if (out == nullptr || out_len == 0)
		return false;

	uint8_t len = last_number_stored_len();
	if (len == 0 || len > (uint8_t)(out_len - 1))
		return false;

	for (uint8_t i = 0; i < len; i++)
		out[i] = (char)EEPROM.read(LAST_NUMBER_OFF_CHARS + i);
	out[len] = '\0';
	return true;
}

File: tools/pulse_monitor/last_number.py
"""Persisted last-dialled number, for redial from an empty dial buffer.

Mirror of last_number.cpp. Kept pure (operates on a byte list) so the record
framing can be tested without EEPROM; the firmware supplies the actual reads
and writes.

The record is deliberately self-describing rather than NUL-terminated.
Erased AVR EEPROM reads as 0xFF in every byte and a freshly zeroed one reads
as 0x00, so a decoder that trusts the stored bytes would hand back a "phone
number" on a board that has never placed a call. The magic byte says a record
was written, and the length byte bounds the read so a corrupt record cannot
return whatever happens to sit after it in EEPROM.
"""

from __future__ import annotations

# Arbitrary but fixed; distinguishes a written record from erased EEPROM.
LAST_NUMBER_MAGIC = 0x4E

# Mirrors DIAL_BUF_LEN - 1 in rusp_firmware.ino. The recalled value is copied
# straight back into dial_buf, so it can never be allowed to exceed it.
LAST_NUMBER_MAX_LEN = 29


def encode_record(number: str | None) -> list[int] | None:
	"""Bytes to write, or None when there is nothing worth storing.

	Returning None for an empty number matters: writing a zero-length
	record would make the recall path report that a number exists and then
	load nothing, which reads to the user as the button doing nothing.
	"""
	if not number:
		return None
	text = number[:LAST_NUMBER_MAX_LEN]
	return [LAST_NUMBER_MAGIC, len(text)] + [ord(c) for c in text]


def decode_record(data: list[int]) -> str | None:
	"""Number held in these EEPROM bytes, or None if absent or corrupt."""
	if len(data) < 2:
		return None
	if data[0] != LAST_NUMBER_MAGIC:
		return None
	length = data[1]
	if length < 1 or length > LAST_NUMBER_MAX_LEN:
		return None
	if length > len(data) - 2:
		return None
	return "".join(chr(b) for b in data[2 : 2 + length])
